### DISCUSSION TITLE
The quest of the perfect LoRA

### DISCUSSION THREAD

**Discussion by dorpxam** [kohya-ss/musubi-tuner#336] [General]
Hi. I'm always in a study phase about the LoRA training for Hunyuan Video model.

# Training resolution
After reading some topics here and there about the resolution choice, I'm still not sure if I'm in the right direction. The official Tencent huggingface talk about some inference resolution settings ([here](https://huggingface.co/tencent/HunyuanVideo)). There's two advice for `540p` and `720p` resolutions:

| Resolution  |      16:9     |   N-pixels   |      4:3     |   N-pixels   |      1:1     |   N-pixels   |
| ----------- | ------------- | ------------ | ------------ | ------------ | ------------ | ------------ |
| 540p        |   960 x 544   |   522,240    |   832 x 624  |   519,168    |   720 x 720  |   518,400    |
| 720p        |  1280 x 720   |   921,600    |  1104 x 832  |   918,528    |   960 x 960  |   921,600    |

As you can see, only the `1280x720` match the `960x960` number of pixels, all others are upper or lower in reference to the `1:1` resolution.

Based on theses information and by extrapolation, I extend this table to the lowest resolutions:

| Resolution  |      16:9     |   N-pixels   |      4:3     |   N-pixels   |      1:1     |   N-pixels   |
| ----------- | ------------- | ------------ | ------------ | ------------ | ------------ | ------------ |
| 360p        |   640 x 360   |   230,400    |   552 x 416  |   229,632    |   480 x 480  |   230,400    |
| 540p        |   960 x 544   |   522,240    |   832 x 624  |   519,168    |   720 x 720  |   518,400    |
| 720p        |  1280 x 720   |   921,600    |  1104 x 832  |   918,528    |   960 x 960  |   921,600    |

There's no problem. `360p` is half `720p` and match the logical. The problem comes with the `256p` resolution. If I refer to [this](https://github.com/Tencent-Hunyuan/HunyuanVideo/issues/125) discussion, the `256p` resolution is:

| Resolution  |      16:9     |   N-pixels   |      4:3     |   N-pixels   |      1:1     |   N-pixels   |
| ----------- | ------------- | ------------ | ------------ | ------------ | ------------ | ------------ |
| 256p        |   342 x 192   |    65,664    |   296 x 222  |    65,712    |   256 x 256  |    65,536    |

While, using the same logical than previous `360p`, `540p` and `720p` resolutions, the `256p` must be:

`(480 / 360) x 256 ~= 341 (341.333...)`, and rounding to `340` we got:

| Resolution  |      16:9     |   N-pixels   |      4:3     |   N-pixels   |      1:1     |   N-pixels   |
| ----------- | ------------- | ------------ | ------------ | ------------ | ------------ | ------------ |
| 256p        |   453 x 255   |   115,515    |   392 x 295  |   115,640    |   340 x 340  |   115,600    |

This is pure speculation of course. Now, the idea is to adjust all the resolution to be sure all is divisible by 8.

| Resolution  |      16:9     |   N-eighth   |      4:3     |   N-eighth   |      1:1     |   N-eighth   |
| ----------- | ------------- | ------------ | ------------ | ------------ | ------------ | ------------ |
| 360p        |   640 x 360   |   _80 x 45_    |   552 x 416  |  _69 x 52_   |   480 x 480  |  _60 x 60_   |
| 540p        |   960 x 544   |   _120 x 68_   |   832 x 624  | _104 x 78_   |   720 x 720  |  _90 x 90_   |
| 720p        |  1280 x 720   |   _160 x 90_   |  1104 x 832  | _138 x 104_  |   960 x 960  | _120 x 120_  |

By adjusting the `256p` resolution to match the same rule, here is the final proposition about the correct resolution buckets for the Hunyuan Video model:

| Resolution  |      16:9     |   N-eighth   |      4:3     |   N-eighth   |      1:1     |   N-eighth   |
| ----------- | ------------- | ------------ | ------------ | ------------ | ------------ | ------------ |
| 256p        |   **440 x 256**   |   _55 x 32_    |   **392 x 288** |  _49 x 36_   |   **336 x 336**  |  _42 x 42_   |
| 360p        |   **640 x 360**   |   _80 x 45_    |   **552 x 416**  |  _69 x 52_   |   **480 x 480**  |  _60 x 60_   |
| 540p        |   **960 x 544**   |   _120 x 68_   |   **832 x 624**  | _104 x 78_   |   **720 x 720**  |  _90 x 90_   |
| 720p        |  **1280 x 720**   |   _160 x 90_   |  **1104 x 832**  | _138 x 104_  |   **960 x 960**  | _120 x 120_  |

# Multiresolution training

From what I understand, the amount of training data is larger at low resolutions, so training on resolutions around `256p` allows for better diversity in inference rendering. This makes sense based on this document:

![395808477-70f3ec92-b2d0-4b9e-b17f-f6cabdd44dae](https://github.com/user-attachments/assets/f544202b-6f46-4803-9d3b-0c0214894fab)

Therefore, it seems to me that it might be beneficial to train your LoRA in multi-resolution in order to benefit from both the diversity of low resolution and the details of a higher resolution. What do you think? Could training the same videos in `256p`, `360p` and `540p` (`720p` is certainly too heavy for my GPU) have a beneficial effect? ​​This avoids repetition because the same images are 'seen' in 3 different sizes.

# Single blocks and Double blocks

By experimenting with training on images for a character LoRA and excluding the blocks:
```
--network_dim 32 ^
--network_alpha 32 ^
--network_args "exclude_patterns=[r'.*single_blocks.*', r'.*double_blocks\.[0-9]\..*']"
```
The result is surprising. The size gain is substantial, going from ~230MB to ~61MB while maintaining a similar rendering and better compatibility with third-party LoRAs.

What about a LoRA character based solely on a video corpus? Is it possible to limit the number of blocks (single or partial, and the same for doubles) while preserving the movement characteristics of the trained character? Is there any documentation on blocks? On their impact on rendering?

What are your thoughts on all this? What are your resolution choices and your experience with excluding blocks? 

Thank you in advance for your comments.
Max.

**Comment by Sarania** [comment]
Heya, I've been trying to suss out the best way to train Hunyuan since it came out and at this point I've got my Wan config dialed in really good but my Hunyuan config still needs a bit of tweaking. That said I do have some thoughts on some of your questions/ideas:

I've experimented a lot with multi resolution training from the beginning because it just made intuitive sense to me, and I can say overall it's beneficial but there are some considerations. Firstly I talk about in https://github.com/kohya-ss/musubi-tuner/discussions/182#discussioncomment-13186947 how you can best choose the resolution/length combinations based on your available VRAM and "framepixels" which is just width * height * frames. You can also see the common resolution buckets I use there, but my datasets are usually varied in aspect ratio so the final training resolutions vary a bit.

Once you've got that, you've gotta consider that since the higher resolution buckets take basically the same amount of compute and VRAM as the lower ones but carry less temporal information, if you are training a temporal concept (e.g. movement based) this will affect how long you need to train but how much varies. In my experience, Wan responds to multires training (and mixed image/video as well) REALLY well but Hunyuan only moderately. Of the three models I've released for Hunyuan, one of them I got the best results with multires training and the other two produced better results with only low res training. Hunyuan has embedded guidance which makes training it a bit more finicky. All of the models I've released for Wan the latest versions were trained multires as that consistently gives me the best results.

Yes limiting the active blocks during training is definitely helpful in improving compatibility with other LoRA especially for Hunyuan, decreasing the Rank also helps - I use 16 usually for both Wan and Hunyuan. Deranking(dropping many/most ranks) after training can further improve compatibility(but that's only anecdotal based on what I've seen others report, I haven't tried it myself yet). Yes it's possible to limit the blocks while training on video, my last Hunyuan character LoRA was trained at 16/16 double blocks only and still picked up her movement characteristics and facial expressions etc as well as several outfits and other things I'd captioned for. It was also mixed image/video. The Rank of a LoRA (network_dim) determines it's size in combination with the base models size. Hunyuan, Wan, also Flux these are BIG models and so high ranks aren't necessary to achieve good learning. As far as further documentation, I haven't seen any. It may be out there but not easy to find - that's unfortunately the type of knowledge people like to keep secret and my own experiments have been limited to training double blocks only.

Lastly I mentioned combined image/video datasets a few times and this generally works well but a consideration is say you have 10 images and 5 videos and for simplicity sake a single video bucket of 480x272@65f and a single image bucket of 1280x720. Even though you have more images than videos, each video is effectively 65 images at that low res so the weight of the videos is gonna be much heavier. If you want the images to count for something, it helps to increase their `num_repeats` - I use between 2-4 depending on the rest of my dataset.  

That's all I've got for ya, cheers!

**Comment by dorpxam** [comment]
Oh. Again, thank you @Sarania for taking the time to reply, it means a lot to me. Your advice is valuable and I take particular interest in it.

I'm currently testing the Hunyuan model mainly because the LoRA produced works perfectly with my Framepack T2V workflow. Since I'm running on a 16GB RTX4080 (can't wait for the NVidia Digits!), I'm very limited for local training and even inference for that matter. I'm not ruling out pure Framepack training and also Wan because I've heard a lot of good feedback about it.

Regarding my ideal resolution tables, I realize that there is no real, precise, ideal size. For example, leaving `bucket_no_upscale = false` changes the config resolutions (latest table of my first message) as follows:

| Resolution  |     config    |    musubi    |  crop/resize? |
| ----------- | ------------- | ------------ | ----------- |
| 256p        |   342 x 192   |   336 x 192  |    width    |
| 360p        |   640 x 360   |   640 x 352  |    height   |
| 540p        |   960 x 544   |   960 x 544  |    none     |

I'll try with your setup, I like the idea of ​​fitting your multiresolutions to a single _megaframepixels_. The idea is to find the ideal values ​​as you say.

I see that you are running at a rank of 16 and an identical alpha. This is interesting, I must admit that having to use LoRA (often of poor quality) which are huge (300MB+, even 1GB for someone), is painful. This is why I like the idea of ​​excluding blocks and here, limiting the rank to 16. I will try with the few datasets that I use for experimentation.

Currently, I use the Prodigy optimizer, with the following configuration:
```
--optimizer_type modules.prodigy.Prodigy
--optimizer_args betas=0.9,0.99 weight_decay=0.01 slice_p=11 decouple=True d_coef=0.5 beta3=None use_bias_correction=True
--learning_rate 1
--lr_scheduler_type torch.optim.lr_scheduler.CosineAnnealingLR
--lr_scheduler_args T_max=XXXXX
```
Where `T_max` (from the scheduler) is set to the number of total steps. It works very well in my opinion, but the problem is that it prohibits the use of `--network_args "loraplus_lr_ratio=4"`. I just can put it at `--network_args "loraplus_lr_ratio=1"`, so without effect I think. I'm going to try the CAME optimizer, and then it will allow me to manipulate the learning rate unlike Prodigy. 

I see you're only training double blocks including 0-9, it's interesting to know that it doesn't affect the movements excessively. I don't know why I used block exclusion `--network_args "exclude_patterns=[r'.*single_blocks.*', r'.*double_blocks\.[0-9]\..*']"`, I'll limit to excluding single blocks only.

Two additional quick questions to make sure I follow your recommendations precisely:

1) What is your `batch_size` in the case of multi-resolution training, knowing that multi-resolution induces parallelism (if I'm not mistaken) even with a `batch_size` of 1, in the log, we find `batch size per device 1, 1, 1` for example. And if you don't use the `batch_size`, do you use `--gradient_accumulation_steps` greater than 1?

2) About image/video training. Do you provide a multi-resolution dataset for both images and videos, or just videos? In your opinion, what is the impact of these two types of training? Do images better capture physical details, especially the face and physical features, and conversely, are videos more useful for anything related to movements (whether subtle: mouth movement for example, or more general: body movement while walking).

Thank you very much for all!
Max

P.S.: I don't forget the idea of ​​LoRA merge but I really want to learn more about it and in particular about the impact of the training layers on the final result in order to avoid making mistakes. I think first to find a way to do multiple inference batches by activating only some layers of the LoRA to test in a systemic way in order to see what is the impact of each layer on the final result. I don't know if it's a good idea, but I must say that 'seeing' what happens is always more meaningful than working on matrices of numbers :) On this subject, a very interesting paper concerning the difference between LoRa training and fine-tuning: [LoRA vs Full Fine-tuning: An Illusion of Equivalence](https://arxiv.org/abs/2410.21228)

**Comment by Sarania** [comment]
Hey I'm rocking an RTX 4070 Ti Super 16GB so I feel the pain there! Actually even obtaining that was a major stretch for me I had to save for many months so I'm grateful to have it. Anyway I have no personal experience with Prodigy optimizer, I've been using CAME pretty exclusively since I started training generative video simply because Adamw8bit has never been my friend and I was looking for an alternative to Adafactor which I had been using but can be a bit unstable. I feel like CAME is a hidden gem - it has the memory efficiency of Adafactor with the accuracy of Adam and has been giving me really good results. LoraPlus is definitely helpful so gaining access to that will be good - I find a value of 4 is excellent for Wan and Hunyuan. It's simply a multiplier applied to the learning rate of LoRA B modules, whereas the base rate is still used for LoRA A. 

1. As far as my batch_size, for VRAM reasons I always have it at 1 for video datasets. For images it depends, if I'm doing mixed image/video I'll have my image batch_size at 1 or 2, but if I'm doing image only I might do 2-4. I probably use 2 most often for images and I pretty much universally do my images at [1280, 720] e.g. model native, I haven't experimented with anything higher.

2. I only provide multires for video, I don't think it's necessary for images since you can just do them at native res easily they take up so much less VRAM. Just like my videos though, I purposely make sure the aspect ratio of my images varies a bit so while the final pixel count will be the same, the actual resolution may vary a bit. I found at the beginning when not doing this if your dataset is heavily biased in a particular aspect ratio, your LoRA will also be biased and struggle to make other aspect ratios. I take advantage of watermarks or other unwanted stuff in the source material to dictate how I crop when preparing my dataset. Edit: Oh and yeah I find images are useful for providing higher detail as well as anchoring certain concepts like locations or outfits whereas video is much better at capturing facial expressions and movements as one would expect, as well as a person's "presence" for lack of a better word. Also including videos versus images only just produces a LoRA that plays better with movement in general. Image only LoRA sometimes suffer from stiff, awkward movements.

Yeah I also get frustrated by other's large LoRA, especially when it's some simple concept but they've gone like Rank 64 and it's nearly a gigabyte! I think many people don't understand what rank/dimension actually is and that it's not a "solo" parameter - that the size of the base model interacts with it to create the final size. Especially if they are used to training SD1.5 or XL or other smaller models, they may have fallen into a habit of larger ranks to get the results they want and now don't realize it's actually killing their LoRA's ability to be used with any other LoRA. Hunyuan especially is super sensitive to this sort of thing. I don't usually limit the blocks when training Wan (I might try at some point) but for Hunyuan it's mandatory if you want compatibility with other LoRA. It's very possible that just "double blocks only" like I do is not the most optimal, it's just all I've experimented with so far! 

I actually thought your resolution breakdown was interesting. I personally haven't experimented to see if some resolutions produce better results than others (ceteris paribus), but it seems possible it could especially if the model used specific buckets during it's original training. The resolutions I chose were just based on the framepixels concept I mentioned which is definitely helpful for managing VRAM concerns. You made quite a deep dive into the logic of that!

A side note is for inference especially I've become quite fond of 4:3 and 3:4 aspect ratios. I just feel like they have a better balance of width to height than do 16:9 and 9:16, at least for the type of content I work with which is largely focused on one or more human subjects. For Hunyuan I often use 832x624 or it's inverse and for Wan, 960x720 or it's inverse. I don't know if the fact that I grew up in the late CRT era plays any role in that preference or not XD

That paper's abstract is interesting, especially about the so called "intruder dimensions." I'll have to read it more in depth. Along those lines those back when I was training Flux, I heard rumors that better results could be obtained by doing a full fine tune and then extracting a LoRA versus training a LoRA directly. I never got around to testing that though because Hunyuan happened, and then Wan, and now it's pretty much all video all the time XD.

**Comment by dorpxam** [comment]
You're right. The resolution is always something cryptic. As I understand, there some limitations not only about the classic (something oldschool like CRT ;)) resolution, but to fit into subdivision. That's why I use a modulo of 8 (purely heuristic I confess it), but looking in the code [here](https://github.com/kohya-ss/musubi-tuner/blob/86fcabb5b42cb32ae8b7934926704e7638f751ce/frame_pack/bucket_tools.py#L21) and [here](https://github.com/kohya-ss/musubi-tuner/blob/86fcabb5b42cb32ae8b7934926704e7638f751ce/dataset/image_video_dataset.py#L370), the resolution is adjusted using a mix of:
- The most close bucket resolution around the 640p reference table
- Adjusting with a closest aspect ratio divisible by 16

That's logical in reference of the previous table:

| Resolution  |     config    | mod 16       |   musubi    | mod 16         | crop/resize dimension |
| ----------- | ------------- | ------------ | ----------- | -------------- | --------------------- |
| 256p        |   342 x 192   | 21.375 x 12  |  336 x 192  |    21 x 12    |  width |
| 360p        |   640 x 360   | 40 x 22.5    |  640 x 352  |    40 x 22   | height |
| 540p        |   960 x 544   | 60 x 34      |   960 x 544  |    60 x 34     | none |

I don't know if it's better or not to let musubi-tuner to adjust the resolution and the impact in the quality of the training. But now it's clear, letting `bucket_no_upscale` to `false`, there's a adjustment of the dataset material to fit the `mod16`. In any case, I prefer to prepare my datasets myself so that they do not require any resampling, even if it is of good quality.

To summarize our discussions for those who will read this:

## Dataset

- Depending on the needs, a mix of photos and videos where:
--- Single resolution for the pictures. 
--- Multiples resolutions for the videos adjusted to fit a global `megaframepixels` limit.
- Keep your resolution to a multiple of 16.
- Adjust the repeat of your single pictures to a decent number, keeping a good balance between the number of frames in the videos and controlling the impact of the images alone.

## LoRA

- Use a `rank` of 16, maximum 32 if you use the block exclusion : `--network_dim 16`
- Keep the `alpha` to the same value of the rank : `--network_alpha 16`
- Use the LoRA+ method with a ratio of 4 : `--network_args "loraplus_lr_ratio=4"`

## Optimizer

- For a automatic learning rate adjustment, choose the Prodigy optimizer
- For a VRAM optimized, choose the CAME optimizer that is similar to AdamW.
- Adjust your learning rate to fit your dataset size if you use CAME or similar. 

If anyone has any good advice or would like to share their experience, they are welcome to do so. This thread is open, and the advice provided here is purely experimental.

Thank you @Sarania , I really appreciate your enthusiasm and willingness to share your advice. I wish you the best.

**Reply by Sarania** [reply]
Hey you're welcome! I'm happy to share the knowledge I've got!
